### ISSUE TITLE
win_lgpo.py slow to execute

### ISSUE THREAD

**Issue by dan-rus** [saltstack/salt#50660]
When using states in win_lgpo.py, it is very slow to complete each check.

If I add some warning messages at key points of the set_ function in the state folder's win_lgpo, I can see the following if I have a two states each with one configuration item:
[WARNING ] 0.0 : Start win_lgpo.py set_ function call
[WARNING ] 0.0529649065866 : About to call lgpo.get_policy_info
[WARNING ] 0.791142923417 : Returned from lgpo.get_policy_info
[WARNING ] 0.842564259418 : About to call lgpo.get
[WARNING ] 3.94409042198 : Returned from lgpo.get
[WARNING ] 3.99761096055 : End win_lgpo.py set_ function call
[WARNING ] 4.05679556171 : Start win_lgpo.py set_ function call
[WARNING ] 4.10796445914 : About to call lgpo.get_policy_info
[WARNING ] 4.80997786822 : Returned from lgpo.get_policy_info
[WARNING ] 4.85807965928 : About to call lgpo.get
[WARNING ] 7.70147202406 : Returned from lgpo.get
[WARNING ] 7.74824216828 : End win_lgpo.py set_ function call

The lgpo.get call takes approximately 3 seconds every time, and the lgpo.get_policy_info call which made for each configuration item in the state takes approximately 0.5 seconds, so you can roughly guess how long a call is going to take.  The end result of it in my use-case, is "Total run time:  201.482 s" for 92 states (not all are done via win_lgpo).

If I memoize \_processPolicyDefinitions (It only appears to be reading and parsing the same file over and over), and rewrite 'lgpo.get' so that it only gets the configuration items I am targetting (Currently it gets all GPO settings even if I am only comparing one), then the runtime comes down to:
[WARNING ] 0.0 : Start win_lgpo.py set_ function call
[WARNING ] 0.048824824996 : About to call lgpo.get_policy_info
[WARNING ] 0.788979609464 : Returned from lgpo.get_policy_info
[WARNING ] 0.836328982218 : About to call lgpo.get
[WARNING ] 0.891362816556 : Returned from lgpo.get
[WARNING ] 0.94043829925 : End win_lgpo.py set_ function call
[WARNING ] 0.988341968335 : Start win_lgpo.py set_ function call
[WARNING ] 1.03744282845 : About to call lgpo.get_policy_info
[WARNING ] 1.08731168951 : Returned from lgpo.get_policy_info
[WARNING ] 1.14569445316 : About to call lgpo.get
[WARNING ] 1.19914064034 : Returned from lgpo.get
[WARNING ] 1.24618548403 : End win_lgpo.py set_ function call
The first get_policy_info takes a while, but is cached thereafter and lgpo.get is much faster.

I could put all of the configuration items into a single state and apply it there, but the speed issue is still present as get_policy_info still takes 0.5 seconds for each configuration item check.

Example state
--------------
An example state that would trigger this behaviour is:
Enforce password history:
    lgpo.set:
        - computer_policy:
            "Enforce password history": 24

Version report
---------------
Salt Version:
           Salt: 2018.3.3

Dependency Versions:
           cffi: 1.10.0
       cherrypy: 10.2.1
       dateutil: 2.6.1
      docker-py: Not Installed
          gitdb: 2.0.5
      gitpython: 2.1.3
          ioflo: Not Installed
         Jinja2: 2.9.6
        libgit2: Not Installed
        libnacl: 1.6.1
       M2Crypto: Not Installed
           Mako: 1.0.6
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: 2.17
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.14 (v2.7.14:84471935ed, Sep 16 2017, 20:25:58) [MSC v.1500 64 bit (AMD64)]
   python-gnupg: 0.4.1
         PyYAML: 3.12
          PyZMQ: 16.0.3
           RAET: Not Installed
          smmap: 2.0.5
        timelib: 0.2.4
        Tornado: 4.5.1
            ZMQ: 4.1.6

System Versions:
           dist:
         locale: cp1252
        machine: AMD64
        release: 2016Server
         system: Windows
        version: 2016Server 10.0.14393  Multiprocessor Free

**Comment by dan-rus** [comment]
I should add that after memoizing and rewriting a bit of lgpo.get, my runtime came down to "Total run time:    2.499 s" for the 92 states.

**Comment by twangboy** [comment]
This was address on the develop branch with https://github.com/saltstack/salt/pull/49919. I'd be interested to see how you did it though.

**Comment by mike2523** [comment]
@dan-rus - Would like to see how you did this as well.

**Comment by dan-rus** [comment]
For memoizing the _processPolicyDefinitions bit, I added the following function to modules/win_lgpo.py

```python
class key_memoized(object):
    def __init__(self, func):
       self.func = func
       self.cache = {}

    def __call__(self, *args, **kwargs):
        key = self.key(args, kwargs)
        if key not in self.cache:
            self.cache[key] = self.func(*args, **kwargs)
        return self.cache[key]

    def normalize_args(self, args, kwargs):
        spec = inspect.getargs(self.func.__code__).args
        return dict(kwargs.items() + zip(spec, args))

    def key(self, args, kwargs):
        a = self.normalize_args(args, kwargs)
        return tuple(sorted(a.items()))
```

and then altered the _processPolicyDefinitions function in modules/win_lgpo.py to:

```python
@key_memoized
def _processPolicyDefinitions(policy_def_path='c:\\Windows\\PolicyDefinitions',
                              display_language='en-US'):
```

For the lgpo.get one, I altered the calls to lgpo.get in states/win_lgpo.py to include the pol_data dict:

```python
    current_policy = __salt__['lgpo.get'](policy_class=policy_class,
                                          adml_language=adml_language,
                                          hierarchical_return=False,
                                          pol_data=pol_data)
```

```python
                ret['changes'] = salt.utils.dictdiffer.deep_diff(
                    current_policy,
                    __salt__['lgpo.get'](policy_class=policy_class,
                                         adml_language=adml_language,
                                         hierarchical_return=False,
                                         pol_data=pol_data))
```

Then updated modules/win_lgpo.py so that the get function accepted a pol_data parameter with None as the default value

```python
def get(policy_class=None, return_full_policy_names=True,
        hierarchical_return=False, adml_language='en-US',
        return_not_configured=False, pol_data=None):
```

Then when it is looping through all of the possible values to set, added an if statement before 'if pol:' to decide if I wanted to get that items current value

```python
            if (pol_data is None) or (_pol['Policy'] in pol_data[p_class.lower()]['policy_lookup']):
                if _pol:
... etc.
```

That way, if something called the function without passing the pol_data parameter, it'd get set to None and populate for every value of _pol, but if I passed it a set of parameters that I was interested in, it'd just grab those.

Glad to hear it was already fixed!  I probably should have checked for that!

**Comment by twangboy** [comment]
I think the develop stuff was backported as well... this issue may be resolved in 2018.3.4. The above PR is some additional optimization.
